engine/runtime/src/main/java/org/enso/interpreter/runtime/SerializationPool.java

```diff
@@ -0,0 +1,187 @@
+package org.enso.interpreter.runtime;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import org.enso.pkg.QualifiedName;
+
+final class SerializationPool {
```

Comment:
`SerializationPool` contains all the _threading logic_ of original `SerializationManager`. It replaces `Thread.sleep` with `Object.wait` and it trades the parallel executor for a `newSingleThreadExecutor`. Otherwise the design with `isSerializing` and `isWaitingForSerialization` remains unchanged.